filters/auth/grantconfig.go

```diff
@@ -0,0 +1,280 @@
+
+	// TokeninfoClient, optional. When set, it will be used for the
+	// authorization requests to TokeninfoURL. When not set, a new default
+	// client is created.
+	TokeninfoClient *authClient
+
+	// AuthClient, optional. When set, it will be used for the
+	// access code exchange requests to TokenURL. When not set, a new default
+	// client is created.
+	AuthClient *net.Client
+
+	// DisableRefresh prevents refreshing the token.
+	DisableRefresh bool
+
+	// AuthURLParameters, optional. Extra URL parameters to add when calling
+	// the OAuth2 authorize or token endpoints.
+	AuthURLParameters map[string]string
+
+	// AccessTokenHeaderName, optional. When set, the access token will be set
+	// on the request to a header with this name.
+	AccessTokenHeaderName string
+
+	// TokeninfoSubjectKey, optional. When set, it is used to look up the subject
+	// ID in the tokeninfo map received from a tokeninfo endpoint request.
+	TokeninfoSubjectKey string
+
+	// TokenCookieName, optional. The name of the cookie used to store the
+	// encrypted access token after a successful token exchange.
+	TokenCookieName string
+
+	// ConnectionTimeout used for tokeninfo endpoint.
+	ConnectionTimeout time.Duration
+
+	// MaxIdleConnectionsPerHost used for tokeninfo endpoint.
+	MaxIdleConnectionsPerHost int
+
+	// Tracer used for tokeninfo endpoint.
+	Tracer opentracing.Tracer
+}
+
+var (
+	ErrMissingClientID        = errors.New("missing client ID")
+	ErrMissingClientSecret    = errors.New("missing client secret")
+	ErrMissingSecretsRegistry = errors.New("missing secrets registry")
+	ErrMissingSecretFile      = errors.New("missing secret file")
+	ErrMissingTokeninfoURL    = errors.New("missing tokeninfo URL")
+	ErrMissingProviderURLs    = errors.New("missing provider URLs")
+)
+
+func (c *OAuthConfig) Init() error {
+	if c.initialized {
+		return nil
+	}
+
+	if c.TokeninfoURL == "" {
+		c.initErr = ErrMissingTokeninfoURL
+		return c.initErr
+	}
+
+	if c.AuthURL == "" || c.TokenURL == "" {
+		c.initErr = ErrMissingProviderURLs
+		return c.initErr
+	}
+
+	if c.Secrets == nil {
+		c.initErr = ErrMissingSecretsRegistry
+		return c.initErr
+	}
+
+	if c.SecretFile == "" {
+		c.initErr = ErrMissingSecretFile
+		return c.initErr
+	}
+
+	if c.ClientID == "" && c.ClientIDFile == "" {
+		c.initErr = ErrMissingClientID
+		return c.initErr
+	}
+
+	if c.ClientSecret == "" && c.ClientSecretFile == "" {
+		c.initErr = ErrMissingClientSecret
+		return c.initErr
+	}
+
+	if c.CallbackPath == "" {
+		c.CallbackPath = defaultCallbackPath
+	}
+
+	if c.TokenCookieName == "" {
+		c.TokenCookieName = defaultTokenCookieName
+	}
+
+	if c.TokeninfoClient == nil {
+		client, err := newAuthClient(
+			c.TokeninfoURL,
+			"granttokeninfo",
+			c.ConnectionTimeout,
+			c.MaxIdleConnectionsPerHost,
+			c.Tracer,
+		)
+		if err != nil {
+			return err
+		}
+		c.TokeninfoClient = client
+	}
+
+	if c.AuthClient == nil {
+		c.AuthClient = net.NewClient(net.Options{})
```

Comment:
We should pass the parameters, that we can here via Options.
For example Tracer, MaxIdleConnectionsPerHost and ConnectionTimeout